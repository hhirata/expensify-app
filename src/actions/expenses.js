import database from '../firebase/firebase'
import expenses from '../reducers/expenses';

export const addExpense = (expense) => ({
    type:'ADD_EXPENSE',
    expense
})

export const startAddExpense = (expenseData = {}) =>{
    return (dispatch) => {
        const {
            description = '',
            note= '',
            amount= 0,
            createdAt = 0
        } = expenseData
        const expense = { description, note , amount, createdAt}
        return database.ref('expenses').push( expense )
        .then((ref)=>{
            dispatch(addExpense({
                id:ref.key,
                ...expense
            }))
        })
        .catch( (e) => {

        })
    }
}
export const removeExpense = ({ id }) => (
    {
        type:'REMOVE_EXPENSE',
        id
    }
)

export const editExpense = (id,update) => (
    {
        type:'EDIT_EXPENSE',
        id,
        update
    }
)

export const setExpenses = (expenses) => (
    {
        type:'SET_EXPENSES',
        expenses
    }
)

export const startSetExpenses = () =>{
    return (dispatch) => {
        return database.ref('expenses').once('value')
                .then((snapshot) => {
                    const expensesFromFirebase =[]
                    snapshot.forEach(( childrenSnapshot) => {
                        expensesFromFirebase.push({
                            id:childrenSnapshot.key,
                            ...childrenSnapshot.val()
                        })
                    })
                   dispatch(setExpenses(expensesFromFirebase))
                })
                .catch((e)=>{})
    }
}


export const startRemoveExpense = ( {id} ) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove()
        .then(()=>{
            dispatch(removeExpense({ id }))
        })
    }
}

export const startEditExpense = (id,update) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).update(update)
                .then(()=>{
                    dispatch(editExpense(id,update))
                })
    }
}