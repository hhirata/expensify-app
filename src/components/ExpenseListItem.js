import React from 'react'
import { Link } from 'react-router-dom'
const ExpenseListItem =  ({ id, description,  amount, createdAt, dispatch }) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>Description: {description}</h3>
        </Link>
        <p>Amount: {amount}</p>
        <p>CreatedAt: {createdAt}</p>

    </div>
)

export default ExpenseListItem