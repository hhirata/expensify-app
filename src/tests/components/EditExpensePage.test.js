import React from 'react'
import {shallow} from 'enzyme'
import {EditExpensePage} from '../../components/EditExpensePage'
import expenses from '../fixtures/expenses'

let editExpense, startRemoveExpense, history, wrapper, expense

beforeEach(()=>{
    editExpense = jest.fn()
    startRemoveExpense = jest.fn()
    history ={
        push:jest.fn()
    }
    expense = expenses[0]
    wrapper = shallow(<EditExpensePage editExpense={editExpense} startRemoveExpense={startRemoveExpense} history={history} expense= {expense} />)
})

test('should render EditExpensePage', ()=>{
    expect(wrapper).toMatchSnapshot()
})

test('should handle editExpense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expense)
    expect(editExpense).toHaveBeenLastCalledWith(expense.id,expense)
    expect(history.push).toHaveBeenLastCalledWith('/')
})

test('should handle removeExpense', () => {
    wrapper.find('button').simulate('click')
    expect(startRemoveExpense).toHaveBeenLastCalledWith({id:expense.id})
    expect(history.push).toHaveBeenLastCalledWith('/')
})