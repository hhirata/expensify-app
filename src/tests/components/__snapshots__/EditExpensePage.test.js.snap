// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 5,
        "createdAt": 0,
        "description": "oi",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
